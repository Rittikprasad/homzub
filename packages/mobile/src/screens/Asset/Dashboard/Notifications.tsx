import React from 'react';
import { StyleSheet, View, TouchableOpacity } from 'react-native';
import { withTranslation, WithTranslation } from 'react-i18next';
import { bindActionCreators, Dispatch } from 'redux';
import { connect } from 'react-redux';
import { AlertHelper } from '@homzhub/common/src/utils/AlertHelper';
import { ErrorUtils } from '@homzhub/common/src/utils/ErrorUtils';
import { CommonParamList } from '@homzhub/mobile/src/navigation/Common';
import { DashboardRepository } from '@homzhub/common/src/domain/repositories/DashboardRepository';
import { NotificationService } from '@homzhub/common/src/services/NotificationService';
import { OfferActions } from '@homzhub/common/src/modules/offers/actions';
import { PortfolioActions } from '@homzhub/common/src/modules/portfolio/actions';
import { SearchActions } from '@homzhub/common/src/modules/search/actions';
import { TicketActions } from '@homzhub/common/src/modules/tickets/actions';
import { theme } from '@homzhub/common/src/styles/theme';
import Icon, { icons } from '@homzhub/common/src/assets/icon';
import { EmptyState } from '@homzhub/common/src/components/atoms/EmptyState';
import { NotificationBox } from '@homzhub/common/src/components/molecules/NotificationBox';
import { SearchBar } from '@homzhub/common/src/components/molecules/SearchBar';
import { UserScreen } from '@homzhub/mobile/src/components/HOC/UserScreen';
import {
  AssetNotifications,
  Notifications as NotificationModel,
} from '@homzhub/common/src/domain/models/AssetNotifications';
import { NotificationType } from '@homzhub/common/src/domain/models/DeeplinkMetaData';
import { IFilter } from '@homzhub/common/src/domain/models/Search';
import { NavigationScreenProps, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';
import { DetailType, ListingType } from '@homzhub/common/src/domain/repositories/interfaces';
import { ICurrentOffer } from '@homzhub/common/src/modules/offers/interfaces';
import { ISetAssetPayload } from '@homzhub/common/src/modules/portfolio/interfaces';
import { ICurrentTicket } from '@homzhub/common/src/modules/tickets/interface';
import { NotificationScreens } from '@homzhub/mobile/src/services/constants';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';

interface IDispatchProps {
  setCurrentAsset: (payload: ISetAssetPayload) => void;
  setFilter: (payload: IFilter) => void;
  setCurrentTicket: (payload: ICurrentTicket) => void;
  setCurrentOfferPayload: (payload: ICurrentOffer) => void;
}

type libraryProps = NavigationScreenProps<CommonParamList, ScreensKeys.AssetNotifications>;
type Props = WithTranslation & libraryProps & IDispatchProps;

interface IAssetNotificationsState {
  notifications: AssetNotifications;
  searchText: string;
  limit: number;
  offset: number;
  scrollEnabled: boolean;
}

export class Notifications extends React.PureComponent<Props, IAssetNotificationsState> {
  public state = {
    notifications: {} as AssetNotifications,
    searchText: '',
    limit: 50,
    offset: 0,
    scrollEnabled: true,
  };

  public componentDidMount = async (): Promise<void> => {
    await this.getAssetNotifications();
  };

  public render(): React.ReactNode {
    const {
      t,
      route: { params },
    } = this.props;
    const { scrollEnabled, notifications, searchText } = this.state;
    const getTitle = (): string => {
      if (params) {
        return params.isFromDashboard ? t('dashboard') : params.screenTitle ?? t('assetMore:more');
      }
      return t('assetMore:more');
    };

    const markAsReadIcon = (): React.ReactElement => {
      const isDisabled = notifications?.unreadCount === 0;
      return (
        <TouchableOpacity disabled={isDisabled} onPress={this.handleMarkAsRead}>
          <Icon name={icons.tickWithCircle} size={20} color={isDisabled ? theme.colors.disabled : theme.colors.blue} />
        </TouchableOpacity>
      );
    };

    return (
      <UserScreen
        isOuterScrollEnabled={scrollEnabled}
        title={getTitle()}
        onBackPress={this.handleIconPress}
        pageTitle={t('notification')}
        rightNode={!searchText ? markAsReadIcon() : undefined}
      >
        {this.renderNotifications()}
      </UserScreen>
    );
  }

  public renderNotifications = (): React.ReactElement => {
    const {
      t,
      route: { params },
    } = this.props;
    const { notifications, searchText } = this.state;
    const showSearch = !params || (params && !params.isFromPortfolio);
    let containerStyle = {
      height: 800,
    };
    if (notifications?.results && notifications?.results.length === 0) {
      containerStyle = {
        height: 200,
      };
    }
    return (
      <View style={[styles.searchBarContainer, containerStyle]}>
        {showSearch && (
          <SearchBar
            placeholder={t('searchByKeyword')}
            value={searchText}
            updateValue={this.onUpdateSearchText}
            containerStyle={styles.searchbar}
          />
        )}
        {notifications?.results && notifications?.results.length === 0 && <EmptyState />}
        {notifications?.results && notifications?.results.length > 0 && (
          <NotificationBox
            data={notifications?.results}
            onPress={this.onNotificationClicked}
            unreadCount={notifications?.unreadCount ?? 0}
            shouldEnableOuterScroll={this.toggleScroll}
            onLoadMore={this.onLoadMore}
          />
        )}
      </View>
    );
  };

  // TODO:(Shikha) - Refactor Notification navigation logic.
  public onNotificationClicked = async (data: NotificationModel): Promise<void> => {
    const { notifications } = this.state;
    const { navigation, setFilter, setCurrentTicket, setCurrentOfferPayload, setCurrentAsset } = this.props;
    const {
      id,
      deeplinkMetadata: {
        objectId,
        type,
        assetId,
        leaseListingId,
        saleListingId,
        screen,
        leaseUnitId,
        messageGroupId,
        imageLink,
      },
      isRead,
    } = data;

    if (!isRead) {
      await DashboardRepository.updateNotificationStatus(id);
      this.setState({ notifications: NotificationService.getUpdatedNotifications(id, notifications) });
    }

    if (type === NotificationType.SITE_VISIT) {
      navigation.navigate(ScreensKeys.PropertyVisits, { visitId: objectId });
    } else if (type === NotificationType.REVIEW_AND_RATING) {
      navigation.navigate(ScreensKeys.PropertyVisits, { reviewVisitId: objectId });
    } else if (type === NotificationType.PROPERTY_DETAIL || type === NotificationType.PROPERTY_PREVIEW) {
      setFilter({ asset_transaction_type: leaseListingId > 0 ? 0 : 1 });

      // @ts-ignore
      navigation.navigate(ScreensKeys.PropertyAssetDescription, {
        propertyTermId: leaseListingId > 0 ? leaseListingId : saleListingId,
        propertyId: assetId,
      });
    } else if (type === NotificationType.SERVICE_TICKET) {
      setCurrentTicket({ ticketId: objectId });
      navigation.navigate(ScreensKeys.ServiceTicketDetail);
    } else if (type === NotificationType.OFFER) {
      setCurrentOfferPayload({
        type: leaseListingId === -1 ? ListingType.SALE_LISTING : ListingType.LEASE_LISTING,
        listingId: leaseListingId === -1 ? saleListingId : leaseListingId,
        ...(Boolean(messageGroupId.length) && { threadId: messageGroupId }),
      });

      if (screen === NotificationScreens.OffersReceived) {
        // @ts-ignore
        navigation.navigate(ScreensKeys.BottomTabs, {
          screen: ScreensKeys.More,
          params: {
            screen: ScreensKeys.PropertyOfferList,
            initial: false,
            params: { isReceivedFlow: true },
          },
        });
        return;
      }
      if (screen === NotificationScreens.OffersMade) {
        // @ts-ignore
        navigation.navigate(ScreensKeys.BottomTabs, {
          screen: ScreensKeys.More,
          params: {
            screen: ScreensKeys.PropertyOfferList,
            initial: false,
            params: { isReceivedFlow: false },
          },
        });
        return;
      }

      if (screen === NotificationScreens.OfferChats) {
        navigation.navigate(ScreensKeys.ChatScreen, { isFromOffers: true });
        return;
      }
      // @ts-ignore
      navigation.navigate(ScreensKeys.BottomTabs, {
        screen: ScreensKeys.More,
        params: {
          screen: ScreensKeys.OfferDetail,
          initial: false,
        },
      });
    } else if (type === NotificationType.ASSET) {
      setCurrentAsset({
        asset_id: assetId,
        listing_id: leaseUnitId,
        assetType: DetailType.LEASE_UNIT,
      });
      // @ts-ignore
      navigation.navigate(ScreensKeys.BottomTabs, {
        screen: ScreensKeys.Portfolio,
        params: {
          screen: ScreensKeys.PropertyDetailScreen,
          initial: false,
        },
      });
    } else if (type === NotificationType.CAMPAIGN) {
      if (imageLink.length <= 0 || imageLink.split('.').reverse()[0] === 'svg') return;

      // @ts-ignore
      navigation.navigate(ScreensKeys.BottomTabs, {
        screen: ScreensKeys.Dashboard,
        params: {
          screen: ScreensKeys.DashboardLandingScreen,
          initial: false,
          params: { imageLink },
        },
      });
    } else if (type === NotificationType.VALUE_ADDED_SERVICE) {
      // @ts-ignore
      navigation.navigate(ScreensKeys.Service);
    } else if (type === NotificationType.ASSET_DOCUMENT) {
      navigation.navigate(ScreensKeys.DocumentScreen, { isFromDashboard: true, propertyId: assetId });
    } else if (type === NotificationType.REFER_AND_EARN) {
      navigation.navigate(ScreensKeys.ReferEarn);
    } else if (type === NotificationType.DUE) {
      // @ts-ignore
      navigation.navigate(ScreensKeys.DuesScreen);
    }
  };

  public onLoadMore = (): void => {
    const { notifications } = this.state;
    if (notifications.results && notifications.results.length !== notifications.count) {
      this.setState({ offset: notifications.results.length }, () => {
        this.getAssetNotifications(true).then();
      });
    }
  };

  public onUpdateSearchText = (value: string): void => {
    // Set offset to 0 everytime user searches for a string
    this.setState({ searchText: value, limit: 50, offset: 0 }, () => {
      this.getAssetNotifications().then();
    });
  };

  private toggleScroll = (scrollEnabled: boolean): void => {
    this.setState({ scrollEnabled });
  };

  public handleIconPress = (): void => {
    const { navigation } = this.props;
    navigation.goBack();
  };

  public handleMarkAsRead = async (): Promise<void> => {
    const { notifications } = this.state;
    const { t } = this.props;
    if (notifications?.results?.length > 0) {
      try {
        const latestCreatedAt = notifications.results[0].createdAt;
        await DashboardRepository.markAllNotificationsRead(latestCreatedAt);
        await this.getAssetNotifications();
        AlertHelper.success({ message: t('assetDashboard:allNotificationsAreRead') });
      } catch (e) {
        AlertHelper.error({ message: ErrorUtils.getErrorMessage(e.details) });
      }
    }
  };

  public getAssetNotifications = async (loadMore = false): Promise<void> => {
    const { searchText, limit, offset, notifications } = this.state;
    const {
      route: { params },
    } = this.props;
    let requestPayload = {
      limit,
      offset,
      ...((!params || (params && !params.isFromPortfolio)) && searchText.length > 0 ? { q: searchText } : {}),
    };

    if (params && !params.isFromDashboard) {
      const { saleListingId, leaseListingId, leaseTransaction, propertyId } = params;
      const isTransaction = leaseTransaction && leaseTransaction > 0;
      const isAsset = !isTransaction && !leaseListingId && !saleListingId;

      requestPayload = {
        ...requestPayload,
        ...(isTransaction && { lease_transaction_id: leaseTransaction }),
        ...(!isTransaction && leaseListingId && { lease_listing_id: leaseListingId }),
        ...(!isTransaction && saleListingId && { sale_listing_id: saleListingId }),
        ...(isAsset && { asset_id: propertyId }),
      };
    }

    try {
      const response = await DashboardRepository.getAssetNotifications(requestPayload);
      if (loadMore) {
        this.setState({
          notifications: NotificationService.transformNotificationsData(response, notifications),
        });
      } else {
        this.setState({
          notifications: response,
        });
      }
    } catch (e) {
      AlertHelper.error({ message: ErrorUtils.getErrorMessage(e.details) });
    }
  };
}

export const mapDispatchToProps = (dispatch: Dispatch): IDispatchProps => {
  const { setCurrentAsset } = PortfolioActions;
  const { setFilter } = SearchActions;
  const { setCurrentTicket } = TicketActions;
  const { setCurrentOfferPayload } = OfferActions;
  return bindActionCreators({ setCurrentAsset, setFilter, setCurrentTicket, setCurrentOfferPayload }, dispatch);
};

export default connect(
  null,
  mapDispatchToProps
)(withTranslation(LocaleConstants.namespacesKey.assetDashboard)(Notifications));

const styles = StyleSheet.create({
  searchbar: {
    margin: theme.layout.screenPadding,
  },
  searchBarContainer: {
    backgroundColor: theme.colors.white,
  },
});
