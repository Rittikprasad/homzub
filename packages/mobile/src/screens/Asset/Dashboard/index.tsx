import React from 'react';
import { StyleProp, ViewStyle } from 'react-native';
import { withTranslation, WithTranslation } from 'react-i18next';
import { bindActionCreators, Dispatch } from 'redux';
import { connect } from 'react-redux';
import { SvgProps } from 'react-native-svg';
import { AlertHelper } from '@homzhub/common/src/utils/AlertHelper';
import { ErrorUtils } from '@homzhub/common/src/utils/ErrorUtils';
import { FunctionUtils } from '@homzhub/common/src/utils/FunctionUtils';
import { AnalyticsService } from '@homzhub/common/src/services/Analytics/AnalyticsService';
import { ObjectMapper } from '@homzhub/common/src/utils/ObjectMapper';
import { EventType } from '@homzhub/common/src/services/Analytics/EventType';
import { AssetRepository } from '@homzhub/common/src/domain/repositories/AssetRepository';
import { DashboardRepository } from '@homzhub/common/src/domain/repositories/DashboardRepository';
import { AssetActions } from '@homzhub/common/src/modules/asset/actions';
import { PortfolioActions } from '@homzhub/common/src/modules/portfolio/actions';
import { RecordAssetActions } from '@homzhub/common/src/modules/recordAsset/actions';
import { UserActions } from '@homzhub/common/src/modules/user/actions';
import { SearchActions } from '@homzhub/common/src/modules/search/actions';
import { CommonSelectors } from '@homzhub/common/src/modules/common/selectors';
import { SearchSelector } from '@homzhub/common/src/modules/search/selectors';
import { UserSelector } from '@homzhub/common/src/modules/user/selectors';
import AddProperty from '@homzhub/common/src/assets/images/addProperty.svg';
import Alert from '@homzhub/common/src/assets/images/alert.svg';
import CustomSearch from '@homzhub/common/src/assets/images/customSearch.svg';
import Finance from '@homzhub/common/src/assets/images/finance.svg';
import Help from '@homzhub/common/src/assets/images/headphone.svg';
import ServiceRequest from '@homzhub/common/src/assets/images/serviceRequest.svg';
import {
  AssetAdvertisementBanner,
  AssetMetricsList,
  AssetSummary,
  FullScreenAssetDetailsCarousel,
} from '@homzhub/mobile/src/components';
import AssetMarketTrends from '@homzhub/mobile/src/components/molecules/AssetMarketTrends';
import IconSheet, { ISheetData } from '@homzhub/mobile/src/components/molecules/IconSheet';
import UserSubscriptionPlan from '@homzhub/common/src/components/molecules/UserSubscriptionPlan';
import WelcomeCard from '@homzhub/common/src/components/molecules/WelcomeCard';
import FinanceOverview from '@homzhub/mobile/src/components/organisms/FinanceOverview';
import LeasePropertyList from '@homzhub/mobile/src/components/organisms/LeasePropertyList';
import PendingPropertyListCard from '@homzhub/mobile/src/components/organisms/PendingPropertyListCard';
import VisitListCard from '@homzhub/mobile/src/components/organisms/VisitListCard';
import { UserScreen } from '@homzhub/mobile/src/components/HOC/UserScreen';
import { Asset, PropertyStatus } from '@homzhub/common/src/domain/models/Asset';
import { Filters } from '@homzhub/common/src/domain/models/AssetFilter';
import { IActions, ISelectedAssetPlan } from '@homzhub/common/src/domain/models/AssetPlan';
import { IMetricsData } from '@homzhub/common/src/domain/models/AssetMetrics';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';
import { NavigationScreenProps, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';
import { DashboardNavigatorParamList } from '@homzhub/mobile/src/navigation/DashboardStack';
import { Attachment } from '@homzhub/common/src/domain/models/Attachment';
import { Currency } from '@homzhub/common/src/domain/models/Currency';
import { IFilter } from '@homzhub/common/src/domain/models/Search';
import { IState } from '@homzhub/common/src/modules/interfaces';
import { ISetAssetPayload } from '@homzhub/common/src/modules/portfolio/interfaces';
import { IAssetVisitPayload, VisitStatus } from '@homzhub/common/src/domain/repositories/interfaces';
import { IApiClientError } from '@homzhub/common/src/network/ApiClientError';

interface IDispatchProps {
  setCurrentFilter: (payload: Filters) => void;
  setAssetId: (payload: number) => void;
  setSelectedPlan: (payload: ISelectedAssetPlan) => void;
  setAddPropertyFlow: (payload: boolean) => void;
  setCurrentAsset: (payload: ISetAssetPayload) => void;
  getAssetVisit: (payload: IAssetVisitPayload) => void;
  setInitialState: () => void;
  getTenanciesDetails: () => void;
  getFilterDetails: (payload: IFilter) => void;
  setFilter: (payload: IFilter) => void;
  setEditPropertyFlow: (payload: boolean) => void;
  toggleEditPropertyFlowBottomSheet: (payload: boolean) => void;
}

interface IReduxStateProps {
  assets: Asset[];
  filters: IFilter;
  defaultCurrency: Currency;
}

type libraryProps = NavigationScreenProps<DashboardNavigatorParamList, ScreensKeys.DashboardLandingScreen>;
type Props = WithTranslation & libraryProps & IDispatchProps & IReduxStateProps;

interface IDashboardState {
  metrics: IMetricsData;
  pendingProperties: Asset[];
  isLoading: boolean;
  showBottomSheet: boolean;
}

enum MetricType {
  OFFER = 'OFFER',
  SITE_VISITS = 'SITE_VISIT',
  SHORTLISTED = 'SHORTLISTED',
}

const ShowInMvpRelease = false;

export class Dashboard extends React.PureComponent<Props, IDashboardState> {
  public focusListener: any;
  public blurListener: any;

  public state = {
    metrics: {} as IMetricsData,
    pendingProperties: [],
    isLoading: false,
    showBottomSheet: false,
  };

  public componentDidMount = (): void => {
    const { navigation } = this.props;
    this.focusListener = navigation.addListener('focus', () => {
      this.getScreenData().then();
    });
    this.blurListener = navigation.addListener('blur', () => {
      this.handleScreenBlur();
    });
  };

  public componentWillUnmount(): void {
    this.focusListener();
    this.blurListener();
  }

  public render = (): React.ReactElement => {
    const { t, route } = this.props;
    const { isLoading, pendingProperties, metrics } = this.state;

    return (
      <>
        <UserScreen loading={isLoading} isGradient title={t('dashboard')} onPlusIconClicked={this.openBottomSheet}>
          {this.renderAssetMetricsAndUpdates()}
          {pendingProperties.length > 0 && (
            <PendingPropertyListCard
              data={pendingProperties}
              onPressComplete={this.onCompleteDetails}
              onSelectAction={this.handleActionSelection}
              onViewProperty={this.onViewProperty}
            />
          )}
          {!metrics?.metricValues?.assets?.count && <WelcomeCard handlePress={this.handleAddProperty} />}
          {metrics?.isTenant && <LeasePropertyList />}
          {metrics?.isTenant && <VisitListCard />}
          {!metrics?.isTenant && <FinanceOverview />}
          <AssetMarketTrends isDashboard onViewAll={this.onViewAll} onTrendPress={this.onTrendPress} />
          <AssetAdvertisementBanner />
          {ShowInMvpRelease && <UserSubscriptionPlan onApiFailure={this.onAssetSubscriptionApiFailure} />}
          {this.renderBottomSheet()}
        </UserScreen>
        {Boolean(route?.params?.imageLink && route?.params?.imageLink?.length > 0) && this.renderFullImageView()}
      </>
    );
  };

  public renderAssetMetricsAndUpdates = (): React.ReactElement => {
    const { t } = this.props;
    const { metrics } = this.state;
    return (
      <>
        <AssetMetricsList
          title={`${metrics?.metricValues?.assets?.count ?? 0}`}
          data={metrics?.metricValues?.miscellaneous ?? []}
          subscription={metrics?.metricValues?.userServicePlan?.label}
          onMetricsClicked={this.handleMetricsNavigation}
          subTitleText={metrics?.isTenant ? t('assetPortfolio:tenancies') : t('common:properties')}
        />
        <AssetSummary
          notification={metrics?.updates?.notifications?.count ?? 0}
          serviceTickets={metrics?.updates?.tickets?.open?.count ?? 0}
          dues={metrics?.updates?.dues?.count ?? 0}
          messages={metrics?.updates?.messages?.unread?.count ?? 0}
          containerStyle={styles.assetCards()}
          onPressServiceTickets={this.handleServiceTickets}
          onPressNotification={this.handleNotification}
          onPressMessages={this.handleMessages}
        />
      </>
    );
  };

  public renderBottomSheet = (): React.ReactElement => {
    const { showBottomSheet } = this.state;
    const data = this.formatBottomSheetData();

    return <IconSheet data={data} onCloseSheet={this.closeBottomSheet} isVisible={showBottomSheet} numOfColumns={3} />;
  };

  public renderFullImageView = (): React.ReactElement | null => {
    const { route } = this.props;

    if (route?.params?.imageLink) {
      const imageLink = ObjectMapper.deserialize(Attachment, { link: route?.params?.imageLink });

      const onPressCross = (): void => {
        this.handleScreenBlur();
      };

      return (
        <>
          <FullScreenAssetDetailsCarousel
            data={[imageLink]}
            activeSlide={0}
            onFullScreenToggle={onPressCross}
            hasOnlyImages
          />
        </>
      );
    }

    return null;
  };

  // HANDLERS
  private onAssetSubscriptionApiFailure = (err: IApiClientError): void => {
    AlertHelper.error({ message: ErrorUtils.getErrorMessage(err) });
  };

  private onCompleteDetails = (assetId: number): void => {
    const { navigation, setAssetId, setEditPropertyFlow, toggleEditPropertyFlowBottomSheet } = this.props;
    setAssetId(assetId);
    setEditPropertyFlow(true);
    toggleEditPropertyFlowBottomSheet(true);
    // @ts-ignore
    navigation.navigate(ScreensKeys.PropertyPostStack, {
      screen: ScreensKeys.AddProperty,
      params: { previousScreen: ScreensKeys.Dashboard },
    });
  };

  private onViewAll = (): void => {
    const { navigation } = this.props;
    navigation.navigate(ScreensKeys.MarketTrends, { isFromDashboard: true });
  };

  private onTrendPress = (url: string, trendId: number): void => {
    const { navigation } = this.props;
    navigation.navigate(ScreensKeys.WebViewScreen, { url, trendId });
  };

  private onViewProperty = (data: ISetAssetPayload): void => {
    const { setCurrentAsset, navigation } = this.props;
    setCurrentAsset(data);
    navigation.navigate(ScreensKeys.PropertyDetailScreen);
  };

  private handleScreenBlur = (): void => {
    const { navigation } = this.props;
    navigation.setParams({ imageLink: undefined });
  };

  private handleMessages = (): void => {
    const { navigation } = this.props;
    navigation.navigate(ScreensKeys.Messages, { isFromDashboard: true });
  };

  private handleServiceTickets = (): void => {
    const { navigation } = this.props;
    navigation.navigate(ScreensKeys.ServiceTicketScreen, { isFromDashboard: true });
  };

  private handleAddProperty = (): void => {
    const { navigation } = this.props;
    // @ts-ignore
    navigation.navigate(ScreensKeys.PropertyPostStack, { screen: ScreensKeys.AssetLocationSearch });
    AnalyticsService.track(EventType.AddPropertyInitiation);
  };

  private handleNotification = (): void => {
    const { navigation } = this.props;
    navigation.navigate(ScreensKeys.AssetNotifications, { isFromDashboard: true });
  };

  private handleMetricsNavigation = (name: string): void => {
    const { navigation, setCurrentFilter, t } = this.props;

    const handleNavigation = (): void => {
      const callNavigation = (screen: ScreensKeys): void =>
        // @ts-ignore
        navigation.navigate(screen, { screenTitle: t('assetDashboard:dashboard') });

      switch (name) {
        case MetricType.OFFER:
          callNavigation(ScreensKeys.PropertyOfferList);
          break;
        case MetricType.SHORTLISTED:
          callNavigation(ScreensKeys.SavedPropertiesScreen);
          break;
        case MetricType.SITE_VISITS:
          callNavigation(ScreensKeys.PropertyVisits);
          break;
        default:
          FunctionUtils.noop();
      }
    };

    if ([MetricType.OFFER, MetricType.SHORTLISTED, MetricType.SITE_VISITS].includes(name as MetricType)) {
      handleNavigation();
      return;
    }
    setCurrentFilter(name as Filters);
    // @ts-ignore
    navigation.navigate(ScreensKeys.Portfolio, {
      screen: ScreensKeys.PortfolioLandingScreen,
      params: {
        isFromNavigation: true,
      },
    });
  };

  private handleActionSelection = (item: IActions, assetId: number): void => {
    const { navigation, setSelectedPlan, setAssetId } = this.props;
    setSelectedPlan({ id: item.id, selectedPlan: item.type });
    setAssetId(assetId);
    // @ts-ignore
    navigation.navigate(ScreensKeys.PropertyPostStack, {
      screen: ScreensKeys.AssetListing,
      params: { previousScreen: ScreensKeys.Dashboard },
    });
  };

  private openBottomSheet = (): void => this.setState({ showBottomSheet: true });

  private closeBottomSheet = (): void => {
    const { setInitialState } = this.props;
    setInitialState();
    this.setState({ showBottomSheet: false });
  };

  private formatBottomSheetData = (): ISheetData[] => {
    const { t, navigation, assets, getFilterDetails, filters, setFilter, defaultCurrency } = this.props;
    const iconSize = 38;

    const handleAddFinancialRecord = (): void => {
      if (assets.length <= 0) {
        AlertHelper.error({ message: t('assetFinancial:addProperty') });
        return;
      }
      navigation.navigate(ScreensKeys.AddRecordScreen, { isFromDashboard: true });
    };

    const handleAddServiceTicket = (): void => {
      if (assets.length <= 0) {
        AlertHelper.error({ message: t('assetDashboard:addPropertyForTicket') });
        return;
      }
      navigation.navigate(ScreensKeys.AddServiceTicket, { isFromDashboard: true });
    };

    const handleAddSupportTicket = (isBugReport?: boolean): void => {
      navigation.navigate(ScreensKeys.SupportScreen, {
        isFromDashboard: true,
        ...(isBugReport && { categoryId: 41 }), // For Bug Report hard-coding the bug category id
      });
    };

    const handleCustomSearch = (): void => {
      getFilterDetails({ asset_group: filters.asset_group });
      setFilter({ currency_code: defaultCurrency.currencyCode });
      navigation.navigate(ScreensKeys.SearchRequirement, { isFromAuth: false });
    };

    const ImageHOC = (Image: React.FC<SvgProps>): React.ReactElement => <Image width={iconSize} height={iconSize} />;

    return [
      {
        icon: ImageHOC(AddProperty),
        label: t('property:addProperty'),
        onPress: this.handleAddProperty,
      },
      {
        icon: ImageHOC(Finance),
        label: t('assetFinancial:addFinancials'),
        onPress: handleAddFinancialRecord,
      },
      {
        icon: ImageHOC(ServiceRequest),
        label: t('assetDashboard:serviceTicket'),
        onPress: handleAddServiceTicket,
      },
      {
        icon: ImageHOC(CustomSearch),
        label: t('propertySearch:shareRequirement'),
        onPress: handleCustomSearch,
      },
      {
        icon: ImageHOC(Help),
        label: t('common:askHelp'),
        onPress: handleAddSupportTicket,
      },
      {
        icon: ImageHOC(Alert),
        label: t('common:bugReport'),
        onPress: (): void => handleAddSupportTicket(true),
      },
    ];
  };

  // HANDLERS end

  // APIs
  private getScreenData = async (): Promise<void> => {
    const { getTenanciesDetails } = this.props;
    await this.getAssetMetrics();
    await this.getPendingProperties();
    this.getVisitData();
    getTenanciesDetails();
  };

  private getAssetMetrics = async (): Promise<void> => {
    this.setState({ isLoading: true });
    try {
      const { metricsData } = await DashboardRepository.getAssetMetrics('v4');
      this.setState({ metrics: metricsData, isLoading: false });
    } catch (e) {
      this.setState({ isLoading: false });
      const error = ErrorUtils.getErrorMessage(e.details);
      AlertHelper.error({ message: error });
    }
  };

  private getPendingProperties = async (): Promise<void> => {
    try {
      const response: Asset[] = await AssetRepository.getPropertiesByStatus(PropertyStatus.PENDING);
      this.setState({ pendingProperties: response });
    } catch (e) {
      AlertHelper.error({ message: ErrorUtils.getErrorMessage(e.details) });
    }
  };

  private getVisitData = (): void => {
    const { getAssetVisit } = this.props;
    getAssetVisit({
      start_date__gte: new Date().toISOString(),
      status__in: `${VisitStatus.APPROVED},${VisitStatus.PENDING}`,
    });
  };
  // APIs end
}

export const mapDispatchToProps = (dispatch: Dispatch): IDispatchProps => {
  const { setCurrentFilter, setCurrentAsset, setInitialState, getTenanciesDetails } = PortfolioActions;
  const { setAddPropertyFlow } = UserActions;
  const { setAssetId, setSelectedPlan, setEditPropertyFlow, toggleEditPropertyFlowBottomSheet } = RecordAssetActions;
  const { getAssetVisit } = AssetActions;
  const { getFilterDetails, setFilter } = SearchActions;
  return bindActionCreators(
    {
      setCurrentFilter,
      setAssetId,
      setSelectedPlan,
      setAddPropertyFlow,
      setCurrentAsset,
      setInitialState,
      getAssetVisit,
      getTenanciesDetails,
      getFilterDetails,
      setFilter,
      setEditPropertyFlow,
      toggleEditPropertyFlowBottomSheet,
    },
    dispatch
  );
};

const mapStateToProps = (state: IState): IReduxStateProps => {
  const { getUserAssets } = UserSelector;
  const { getFilters } = SearchSelector;
  const { getDefaultCurrency } = CommonSelectors;
  return {
    assets: getUserAssets(state),
    filters: getFilters(state),
    defaultCurrency: getDefaultCurrency(state),
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withTranslation(LocaleConstants.namespacesKey.assetDashboard)(Dashboard));

const styles = {
  assetCards: (): StyleProp<ViewStyle> => ({
    marginVertical: 12,
  }),
  evenItem: (): StyleProp<ViewStyle> => ({
    marginEnd: 17,
  }),
};
