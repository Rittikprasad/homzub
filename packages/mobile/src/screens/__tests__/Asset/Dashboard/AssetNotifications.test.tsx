import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Notifications } from '@homzhub/mobile/src/screens/Asset/Dashboard/Notifications';

describe('Asset Notifications Screen', () => {
  let component: any;
  let props: any;

  beforeEach(() => {
    props = {
      navigation: {
        goBack: jest.fn(),
      },
      route: {
        params: {
          isFromDashboard: true,
        },
      },
    };
    component = shallow(<Notifications {...props} t={(key: string): string => key} />);
  });

  it('should render snapshot', () => {
    component.setState({ notifications: { results: [{}] } });
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render snapshot with empty notifications', () => {
    component.setState({ notifications: { results: [] } });
    expect(toJson(component)).toMatchSnapshot();
  });
});
