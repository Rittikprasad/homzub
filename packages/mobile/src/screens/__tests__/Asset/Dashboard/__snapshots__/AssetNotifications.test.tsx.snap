// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Asset Notifications Screen should render snapshot 1`] = `
<Memo(UserScreen)
  isOuterScrollEnabled={true}
  onBackPress={[Function]}
  pageTitle="notification"
  rightNode={
    <ForwardRef
      disabled={false}
      onPress={[Function]}
    >
      <Icon
        allowFontScaling={false}
        color="#0084F8"
        name="tick-with-circle"
        size={20}
      />
    </ForwardRef>
  }
  title="dashboard"
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#FFFFFF",
        },
        Object {
          "height": 800,
        },
      ]
    }
  >
    <SearchBar
      containerStyle={
        Object {
          "margin": 16,
        }
      }
      placeholder="searchByKeyword"
      updateValue={[Function]}
      value=""
    />
    <Memo(NotificationBox)
      data={
        Array [
          Object {},
        ]
      }
      onLoadMore={[Function]}
      onPress={[Function]}
      shouldEnableOuterScroll={[Function]}
      unreadCount={0}
    />
  </View>
</Memo(UserScreen)>
`;

exports[`Asset Notifications Screen should render snapshot with empty notifications 1`] = `
<Memo(UserScreen)
  isOuterScrollEnabled={true}
  onBackPress={[Function]}
  pageTitle="notification"
  rightNode={
    <ForwardRef
      disabled={false}
      onPress={[Function]}
    >
      <Icon
        allowFontScaling={false}
        color="#0084F8"
        name="tick-with-circle"
        size={20}
      />
    </ForwardRef>
  }
  title="dashboard"
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#FFFFFF",
        },
        Object {
          "height": 200,
        },
      ]
    }
  >
    <SearchBar
      containerStyle={
        Object {
          "margin": 16,
        }
      }
      placeholder="searchByKeyword"
      updateValue={[Function]}
      value=""
    />
    <EmptyState />
  </View>
</Memo(UserScreen)>
`;
